lib/defender/aws_iot_defender.c

```diff
@@ -0,0 +1,183 @@
+
+/* Period between reports in milliseconds. */
+static uint32_t _periodMilliSecond = AWS_IOT_DEFENDER_DEFAULT_PERIOD_SECONDS * 1000;
+
+/* the timer that connect, subscribe, publish metrics report */
+static AwsIotTimer_t _metricsPublishTimer;
+
+/* the timer that disconnect */
+static AwsIotTimer_t _guardTimer;
+
+static AwsIotMqttConnection_t _mqttConnection = AWS_IOT_MQTT_CONNECTION_INITIALIZER;
+
+static AwsIotDefenderStartInfo_t _startInfo = { 0 };
+
+/**
+ * Return the state that whether defender is already started.
+ */
+static bool isStarted();
+
+/**
+ * Create and arm _metricsPublishTimer
+ */
+static void _metricsPublishTimerExpirationRoutine( void * );
+static void _guardTimerExpirationRoutine( void * );
+
+/*-----------------------------------------------------------*/
+
+static bool isStarted()
+{
+    return _startInfo.pThingName != NULL && _startInfo.thingNameLength > 0;
+}
+
+/*-----------------------------------------------------------*/
+
+AwsIotDefenderError_t AwsIotDefender_Start( AwsIotDefenderStartInfo_t * pStartInfo )
+{
+    AwsIotDefenderError_t defenderError = AWS_IOT_DEFENDER_INTERNAL_FAILURE;
+
+    /* if current state is not started */
+    if( !isStarted() )
+    {
+        /* copy input start info into global variable _startInfo */
+        _startInfo = *pStartInfo;
+
+        /* metrics publish timer must be created and armed */
+        /* guard timer must be created */
+        if( AwsIotClock_TimerCreate( &_metricsPublishTimer, _metricsPublishTimerExpirationRoutine, NULL ) )
+        {
+            if( AwsIotClock_TimerArm( &_metricsPublishTimer, _periodMilliSecond, _periodMilliSecond ) )
+            {
+                if( AwsIotClock_TimerCreate( &_guardTimer, _guardTimerExpirationRoutine, NULL ) )
+                {
+                    defenderError = AWS_IOT_DEFENDER_SUCCESS;
+                }
+            }
+            else
+            {
+                AwsIotClock_TimerDestroy( &_metricsPublishTimer );
+            }
+        }
+    }
+    else
+    {
+        defenderError = AWS_IOT_DEFENDER_ALREADY_STARTED;
+    }
+
+    if( defenderError != AWS_IOT_DEFENDER_SUCCESS )
+    {
+        _startInfo = _START_INFO_RESETER;
+    }
+
+    return defenderError;
+}
+
+/*-----------------------------------------------------------*/
+
+AwsIotDefenderError_t AwsIotDefender_Stop( void )
+{
+    AwsIotClock_TimerDestroy( &_metricsPublishTimer );
+    AwsIotClock_TimerDestroy( &_guardTimer );
+    _startInfo = _START_INFO_RESETER;
+
+    return AWS_IOT_DEFENDER_SUCCESS;
+}
+
+/*-----------------------------------------------------------*/
+
+AwsIotDefenderError_t AwsIotDefender_SetPeriod( uint64_t periodSeconds )
+{
+    AwsIotDefenderError_t defenderError = AWS_IOT_DEFENDER_INTERNAL_FAILURE;
+
+    if( periodSeconds < AWS_IOT_DEFENDER_DEFAULT_PERIOD_SECONDS )
+    {
+        defenderError = AWS_IOT_DEFENDER_PERIOD_TOO_SHORT;
+    }
+    else
+    {
+        _periodMilliSecond = periodSeconds * 1000;
```

Comment:
Check for overflow, although unlikely...

Reply:
Yes, I will add an upper limit